Automattic/studio
# Studio: export progress UI

Related to https://github.com/Automattic/dotcom-forge/issues/8365.

## Proposed Changes

This PR adds an export progress bar and notification.

**Changes:**
- Move part of the export logic to a separate hook (`useImportExport`). In a next PR, I'll also move part of the import logic.
- Add site ID argument to export events.
- Remove export validation events since we no longer emit them after https://github.com/Automattic/studio/pull/391.

<details><summary><strong>Notes</strong></summary>

1. Currently only export progress bar is implemented and possibly we need to make minor UI changes.
2. From time to time ( when you export multiple times in a row ), I see the following error:
![2024-07-22T18:09:06,691314264+03:00](https://github.com/user-attachments/assets/0564a13d-67c1-4870-8ebd-43287ca6d6c7)
We need to address that, either here, or in https://github.com/Automattic/studio/pull/370


</details> 

## Testing Instructions

### Export completes

- Apply the following patch to delay progress steps in the export process:
```patch
diff --git forkSrcPrefix/src/lib/import-export/handle-events.ts forkDstPrefix/src/lib/import-export/handle-events.ts
index 33633d3d5933ca19517922c812fef68735c75dbe..59e360d1a76fc44c6956d6a0c766786aa94f85fb 100644
--- forkSrcPrefix/src/lib/import-export/handle-events.ts
+++ forkDstPrefix/src/lib/import-export/handle-events.ts
@@ -1,5 +1,5 @@
 import { EventEmitter } from 'events';
-import { ExportEventType } from './export/events';
+import { ExportEventType, ExportEvents } from './export/events';
 import { ImportEventType } from './import/events';
 
 export type ImportExportEventType = ImportEventType | ExportEventType;
@@ -9,18 +9,37 @@ export interface ImportExportEventData {
 	data: unknown;
 }
 
+const DELAY_EVENTS: string[] = [
+	ExportEvents.EXPORT_START,
+	ExportEvents.BACKUP_CREATE_START,
+	ExportEvents.BACKUP_CREATE_PROGRESS,
+	ExportEvents.BACKUP_CREATE_COMPLETE,
+	ExportEvents.EXPORT_COMPLETE,
+];
+
 export const handleEvents = (
 	emitter: Partial< EventEmitter >,
 	onEvent: ( data: ImportExportEventData ) => void,
 	events: Record< string, string >
 ) => {
 	const removeListeners: ( () => void )[] = [];
+	let delay = 0;
 	Object.values( events ).forEach( ( eventName ) => {
 		const listener = ( data: unknown ) => {
-			onEvent( {
-				event: eventName as ImportExportEventType,
-				data,
-			} );
+			if ( DELAY_EVENTS.includes( eventName ) ) {
+				setTimeout( () => {
+					onEvent( {
+						event: eventName as ImportExportEventType,
+						data,
+					} );
+				}, delay );
+				delay += eventName.endsWith( '_progress' ) ? 10 : 1000;
+			} else {
+				onEvent( {
+					event: eventName as ImportExportEventType,
+					data,
+				} );
+			}
 		};
 		emitter.on?.( eventName, listener );
 		removeListeners.push( () => emitter.off?.( eventName, listener ) );
diff --git forkSrcPrefix/src/ipc-handlers.ts forkDstPrefix/src/ipc-handlers.ts
index 0b8bf8d36c49090ae3e7a1e64b0c08c899c5e337..e11767c723b1ccb5914ba95a073bfff286643492 100644
--- forkSrcPrefix/src/ipc-handlers.ts
+++ forkDstPrefix/src/ipc-handlers.ts
@@ -20,6 +20,7 @@ import { downloadSqliteIntegrationPlugin } from '../vendor/wp-now/src/download';
 import { LIMIT_ARCHIVE_SIZE } from './constants';
 import { isEmptyDir, pathExists, isWordPressDirectory, sanitizeFolderName } from './lib/fs-utils';
 import { getImageData } from './lib/get-image-data';
+import { ExportEvents } from './lib/import-export/export/events';
 import { exportBackup } from './lib/import-export/export/export-manager';
 import { ExportOptions } from './lib/import-export/export/types';
 import { ImportExportEventData } from './lib/import-export/handle-events';
@@ -484,13 +485,18 @@ export async function exportSite(
 	siteId: string
 ): Promise< void > {
 	try {
+		let onResolve: () => void;
 		const onEvent = ( data: ImportExportEventData ) => {
 			const parentWindow = BrowserWindow.fromWebContents( event.sender );
 			if ( parentWindow && ! parentWindow.isDestroyed() && ! event.sender.isDestroyed() ) {
+				if ( data.event === ExportEvents.EXPORT_COMPLETE ) {
+					onResolve();
+				}
 				parentWindow.webContents.send( 'on-export', data, siteId );
 			}
 		};
 		await exportBackup( options, onEvent );
+		await new Promise< void >( ( resolve ) => ( onResolve = resolve ) );
 	} catch ( e ) {
 		Sentry.captureException( e );
 		throw e;

```
- Select a site.
- Navigate to the Import/Export tab.
- Click on Export entire site.
- Select a destination for the export and click on Save.
- Observe a progress bar is shown with the progress of the export process.
- Select another site.
- Click on Export entire site.
- Select a destination for the export and click on Save.
- Observe a progress bar is shown with the progress of the export process.
- Switch back and forth between sites and observe the progress for each site.
- When the export completes, observe a notification is displayed.

### Encountering an error



- Apply the following patch to force an error:
```patch
diff --git a/src/lib/import-export/export/exporters/default-exporter.ts b/src/lib/import-export/export/exporters/default-exporter.ts
index 661cf3d..741951b 100644
--- a/src/lib/import-export/export/exporters/default-exporter.ts
+++ b/src/lib/import-export/export/exporters/default-exporter.ts
@@ -63,6 +63,7 @@ export class DefaultExporter extends EventEmitter implements Exporter {
 		this.archive.pipe( output );
 
 		try {
+			throw Error( 'test' );
 			this.addWpConfig();
 			this.addWpContent();
 			await this.addDatabase();

```
- Select a site.
- Navigate to the Import/Export tab.
- Click on Export entire site.
- Select a destination for the export and click on Save.
- Observe a progress bar is shown with the progress of the export process.
- Observe an error message is displayed.
- Click on the Ok button.
- Observe the progress bar is no longer shown and the export section gets back to the original state.

## Pre-merge Checklist
